relay-protocol/src/value.rs

```diff
@@ -334,3 +335,172 @@ where
 {
     serde_json::to_value(value).map(Value::from_json)
 }
+
+/// Borrowed version of [`Array`].
+#[derive(Debug, Clone, Copy)]
+pub struct Arr<'a> {
+    _phantom: std::marker::PhantomData<&'a ()>,
+}
+
+/// Borrowed version of [`Object`].
+#[derive(Debug, Clone, Copy)]
+pub struct Obj<'a> {
+    _phantom: std::marker::PhantomData<&'a ()>,
```

Comment:
This phantom causes a change in behavior that I'm not 100% sure about. E.g.  for a span `{"data": {"foo": {"bar": 1}}}`, `span.data.foo` would give you an `Object { "bar": Number(1) }`, now it returns this opaque type. Not sure if this has any practical consequences.

We could instead return `None` for object values, to make it clear that getters for objects and arrays are not yet implemented.

Reply:
We couldn't extract such a value before anyway, since the only data types we could work with were the primitive ones. You can see evidence for that because now you cannot access array or object types anymore, but no code broke for this. 

For metric extraction from transactions, we will have to implement support for arrays, but that wasn't there before either.